### PR TITLE
[receiver/snmp] Add start timestamp to sum metrics

### LINKED ISSUE #17984: [reciever/snmp] Doesn't set StartTimestamp
### Component(s)

receiver/snmp

### What happened?

Working on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17983, I found that the snmp receiver doesn't set StartTimestamp on metrics, at least in the tests. Not sure if it's the actual behavior. While gauges with no StartTimestamp may be okay, Sum metrics with no StartTimestamp are not valid.

### Collector version

latest

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by djaglowski**
Resolves open-telemetry/opentelemetry-collector-contrib#17984

### CODE REVIEW

File: receiver/snmpreceiver/otel_metric_helper.go [hunk 1]
```diff
@@ -87,19 +87,22 @@ type otelMetricHelper struct {
 	// This is the ResourceMetricsSlice that will contain all newly created resources and metrics
 	resourceMetricsSlice pmetric.ResourceMetricsSlice
 	// This is the timestamp that should be added to all created data points
```
Comment (dmitryax) [inline receiver/snmpreceiver/otel_metric_helper.go:89]:
```suggestion
	// This is the start timestamp that should be added to all created data points
```

File: receiver/snmpreceiver/otel_metric_helper.go [hunk 2]
```diff
@@ -174,19 +177,20 @@ func (h *otelMetricHelper) addMetricDataPoint(resourceKey string, metricName str
...
+	var dp pmetric.NumberDataPoint
 	var valueType string
 	if metricCfg.Gauge != nil {
-		dps = metric.Gauge().DataPoints()
+		dp = metric.Gauge().DataPoints().AppendEmpty()
 		valueType = metricCfg.Gauge.ValueType
 	} else {
-		dps = metric.Sum().DataPoints()
+		dp = metric.Sum().DataPoints().AppendEmpty()
+		dp.SetStartTimestamp(h.dataPointStartTime)
```
Comment (dmitryax) [inline receiver/snmpreceiver/otel_metric_helper.go:187]: We can use static start time for cumulative sums only. For deltas, it should be a timestamp of the previous scrape. Can the sums have delta temporality here?

Comment (djaglowski) [reply receiver/snmpreceiver/otel_metric_helper.go:187]: I wouldn't guess it's very common but yes, the user can configure it that way.

Are we handling this in mdatagen? I think we might be making the same mistake [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7c16b41254a80608c7d962674f7cda9754471dd8/cmd/mdatagen/templates/metrics.go.tmpl#L374).

Just spitballing, but I think a proper solution requires a cache that can recall the most recent start time for each data point. It'd probably need to manage values be hierarchically (e.g. resource attributes -> scope attributes -> metric name -> data point attributes) and have a reasonable eviction strategy, possibly configurable. Seems like something worth standardizing. WDYT?

Comment (dmitryax) [reply receiver/snmpreceiver/otel_metric_helper.go:187]: > Are we handling this in mdatagen? I think we might be making the same mistake [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7c16b41254a80608c7d962674f7cda9754471dd8/cmd/mdatagen/templates/metrics.go.tmpl#L374).

> Just spitballing, but I think a proper solution requires a cache that can recall the most recent start time for each data point. It'd probably need to manage values hierarchically (e.g. resource attributes -> scope attributes -> metric name -> data point attributes) and have a reasonable eviction strategy, possibly configurable. Seems like something worth standardizing. WDYT?

That's correct. It's not that simple. I was going to fix the issue in the metrics builder by applying what you described in the scope of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18062

We can leave it as is for now, but maybe capture it in a separate issue?

**Review by dmitryax — APPROVED** [review]

File: receiver/snmpreceiver/otel_metric_helper.go (hunk 2 above)
Comment (djaglowski) [reply receiver/snmpreceiver/otel_metric_helper.go:187]: Ok, makes sense. I've captured it for this receiver here: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18064

**Pull request merged**